File: src/kiro_crew/platform/app_update_bridge.py
"""Gateway <-> packaged-desktop-host bridge for the app update lane.

An agent's update request originates in the Python gateway. On a packaged
install (``dmg``/``appimage``/``deb``/``rpm``) the bytes are replaced by the
Electron main process's updater, which lives in a DIFFERENT process that owns
``autoUpdater`` and the graceful ``stopGateway()`` handoff. So a request raised
on the Python side has to cross a process boundary to reach the only thing that
can act on it.

The direction of control is the same one ``browser/command_bus.py`` already
establishes, and for the same reason: rather than the gateway reaching INTO
Electron (it holds no handle on that process, and opening an inbound control
port on the process that owns the dashboard session is the exposure the browser
channel's design spike already ruled out), the gateway exposes a loopback queue
and **Electron pulls from it**:

* Electron polls ``POST /api/update/app-bridge`` (internal-secret only),
  reporting its updater's discovered state and draining at most one install
  request.
* The gateway records that state — it is the only source of the packaged lane's
  available version, since the CLI release feed describes a different release
  stream — and hands back an install request when a human has approved one.

What this module deliberately does NOT hold is authority. An entry here is a
*request*, produced only by ``POST /api/update/approve`` after the step-up
nonce was consumed (see :mod:`kiro_crew.platform.update_stepup`). Arming is an
agent-reachable action; approving is not. So the queue is only ever written on
the far side of a human approval, and a packaged host that drains it is acting
on a decision the gateway already verified.

State is in-memory and bounded by construction: exactly one host state record
and at most one outstanding install request. A gateway restart discards both,
which is the safe direction — a forgotten request must not survive to install
later, and the host re-reports its state on its next poll.
"""

from __future__ import annotations

import logging
import threading
import time
from dataclasses import dataclass
from typing import Callable

logger = logging.getLogger(__name__)

#: How long a reported host state stays current. A packaged host polls every
#: ``APP_BRIDGE_POLL_MS`` (see ``website/electron/app-update-bridge.js``); this
#: is comfortably above that so one dropped poll does not read as "no desktop
#: host is present", while a host that has actually gone away stops answering
#: for the arm endpoint within a bounded window.
HOST_TTL_S = 120.0

#: How long an approved-but-undrained install request stays deliverable. The
#: step-up TTL already bounds how long an ARM is approvable; this bounds the
#: much shorter window between an approval landing and the host's next poll. A
#: request that ages out is dropped rather than delivered late, because "the
#: human approved this two hours ago" is not consent to quit the app now.
INSTALL_TTL_S = 300.0


@dataclass(frozen=True)
class AppHostState:
    """What a packaged desktop host last reported about its own updater."""

    #: The running app version, as the shell stamps it.
    version: str
    #: The version its updater found on the feed, or ``""`` when it found none.
    available_version: str
    #: The release lane that updater is following.
    channel: str
    #: Monotonic timestamp of the report.
    reported_at: float


@dataclass(frozen=True)
class InstallRequest:
    """One approved install, waiting for the host's next poll."""

    version: str
    channel: str
    request_id: str
    created_at: float


class AppUpdateBridge:
    """Single host state + at most one pending install request.

    Guarded by a plain :class:`threading.Lock` rather than an asyncio lock: the
    handlers that touch it are short and synchronous, and the same state is read
    from executor threads (the arm endpoint's off-loop capability probe runs
    beside it). ``now`` is injectable so the TTLs are testable without sleeping.
    """

    def __init__(self, *, now: Callable[[], float] | None = None) -> None:
        self._now = now or time.monotonic
        self._lock = threading.Lock()
        self._state: AppHostState | None = None
        self._pending: InstallRequest | None = None

    # ── host state ───────────────────────────────────────────────────────────

    def report_state(self, *, version: str, available_version: str, channel: str) -> AppHostState:
        """Record what the packaged host's updater currently knows."""
        state = AppHostState(
            version=str(version or ""),
            available_version=str(available_version or ""),
            channel=str(channel or ""),
            reported_at=self._now(),
        )
        with self._lock:
            self._state = state
        return state

    def host_state(self) -> AppHostState | None:
        """The current report, or ``None`` when absent or stale.

        Staleness is not an error to surface: a host that stopped polling is
        indistinguishable from one that was never there, and both mean the same
        thing to a caller — there is nothing on the far side to act on a
        request, so do not accept one.
        """
        with self._lock:
            state = self._state
            if state is None:
                return None
            if self._now() - state.reported_at > HOST_TTL_S:
                return None
            return state

    def host_present(self) -> bool:
        return self.host_state() is not None

    # ── install requests ─────────────────────────────────────────────────────

    def request_install(self, *, version: str, channel: str, request_id: str) -> InstallRequest:
        """Queue one approved install. Last writer wins.

        Overwriting is correct rather than merely tolerable: each request is the
        product of its own consumed single-use nonce, so a second approval is a
        second human decision, and the newer one describes what the human most
        recently agreed to. Nothing is lost that an install could have honored —
        only one bundle swap can happen.
        """
        req = InstallRequest(
            version=str(version or ""),
            channel=str(channel or ""),
            request_id=str(request_id or ""),
            created_at=self._now(),
        )
        with self._lock:
            self._pending = req
        logger.info(
            "Queued packaged-app install request %s (v%s, %s channel)",
            req.request_id,
            req.version,
            req.channel,
        )
        return req

    def take_install(self) -> InstallRequest | None:
        """Pop the pending request, or ``None``. Single delivery.

        Popping BEFORE the host acts is deliberate: an install quits the app, so
        a request that is delivered and then lost to a crash must not be
        redelivered on the next launch, where the human's approval is long past
        and the quit would be unexplained. Re-approving is cheap; a surprise
        restart is not.
        """
        with self._lock:
            req = self._pending
            self._pending = None
        if req is None:
            return None
        if self._now() - req.created_at > INSTALL_TTL_S:
            logger.info(
                "Dropping packaged-app install request %s: approved %.0fs ago, past the "
                "%.0fs delivery window",
                req.request_id,
                self._now() - req.created_at,
                INSTALL_TTL_S,
            )
            return None
        return req

    def pending_install(self) -> InstallRequest | None:
        """Peek without consuming. For status surfaces and tests only."""
        with self._lock:
            req = self._pending
        if req is None or self._now() - req.created_at > INSTALL_TTL_S:
            return None
        return req

    def clear(self) -> None:
        """Drop all state. Used by tests and by an explicit arm cancellation."""
        with self._lock:
            self._state = None
            self._pending = None


_bridge: AppUpdateBridge | None = None
_bridge_lock = threading.Lock()


def get_app_update_bridge() -> AppUpdateBridge:
    """The process-wide bridge."""
    global _bridge
    with _bridge_lock:
        if _bridge is None:
            _bridge = AppUpdateBridge()
        return _bridge


__all__ = [
    "HOST_TTL_S",
    "INSTALL_TTL_S",
    "AppHostState",
    "AppUpdateBridge",
    "InstallRequest",
    "get_app_update_bridge",
]

File: website/electron/app-update-bridge.js
// Agent -> Electron app-update channel (the CLIENT side).
//
// An agent asking for the desktop app to update raises that request in the
// Python gateway, where the agent surface lives. But the bytes belong to THIS
// process: `autoUpdater`, the staged bundle and the awaited `stopGateway()`
// handoff only exist here. So the request has to cross the process boundary to
// reach the only thing that can act on it.
//
// The direction of control is the same one browser-agent-channel.js already
// established, and for the same reason — the gateway holds no handle on this
// process, and opening an inbound control port on the process that owns the
// dashboard session is exactly the exposure that design ruled out. So the
// gateway exposes a loopback queue and Electron PULLS from it:
//
//   • Electron polls  POST /api/update/app-bridge  with its updater's current
//     state (running version, the version its feed offers, the lane it follows).
//   • That state is the gateway's ONLY source of the packaged lane's available
//     version — the CLI release feed the gateway reads describes a different
//     release stream — so reporting it is what lets `POST /api/update/arm` name
//     a real target instead of refusing every packaged install.
//   • The same answer carries at most one APPROVED install request, which this
//     loop hands to the updater.
//
// ── What arriving here does and does not authorize ───────────────────────────
//
// An install request is only ever produced on the far side of a consumed
// step-up nonce: an agent can ARM an update, and only a caller that read the
// gateway host's trust-fenced nonce file can approve one. So this loop does not
// decide anything — it delivers a decision that was already made and audited,
// and it drives the SAME path the human's Settings > About button drives. It
// does not enable install-on-quit (`autoInstallOnAppQuit` stays false; see
// auto-update.js) and it never calls `quitAndInstall` directly.
//
// ── Why the loop must be defensive ──────────────────────────────────────────
//
// It runs for the life of the app, so the three failure modes the browser
// channel documents apply unchanged: a throwing install dispatch must not kill
// it, a transport error or non-2xx must back off rather than tight-spin, and a
// malformed body must be reported and skipped. Beyond those, ONE more matters
// here: an install request must be dispatched at most once per delivery. The
// gateway pops it, so a loop that re-dispatched on its own would quit the app
// twice for one approval.
//
// ── Testability ─────────────────────────────────────────────────────────────
//
// Every dependency is injected, mirroring browser-agent-channel.js, so the loop
// is unit-testable with a fake `fetchFn`, no Electron, and no real timers:
//
//   fetchFn(url, { method, headers, body }) -> { ok, status, async json() }
//   getGatewayUrl()  -> loopback gateway base URL
//   getSecret()      -> X-Internal-Secret value (NEVER read a file here)
//   getUpdaterState()-> { version, available_version, channel }
//   installApproved({ version, channel, request_id }) -> Promise
//   isGatewayLocal() -> is the gateway on THIS machine?
//   onError(err, ctx)-> log sink; must never throw meaningfully
"use strict";

const BRIDGE_PATH = "/api/update/app-bridge";

// Poll interval (ms). The gateway holds no long-poll here: an approval is a
// human action whose latency budget is seconds, not milliseconds, and a short
// fixed poll keeps the reported updater state fresh without holding a request
// open for the life of the app. Must stay well below the gateway's HOST_TTL_S
// (app_update_bridge.py) so one dropped poll never reads as "no host present".
const DEFAULT_POLL_MS = 30000;
const DEFAULT_BACKOFF_MS = 5000;

/** Join a base URL and a path without doubling or dropping the slash. */
function joinUrl(base, path) {
  const b = String(base || "").replace(/\/+$/, "");
  return `${b}${path}`;
}

/**
 * Normalise an updater state reading into the three strings the gateway takes.
 *
 * A throwing or malformed reader must still produce a report: the host-presence
 * signal is what keeps the arm endpoint answering "the app is reachable", and
 * losing it because a version field came back undefined would silently take the
 * whole feature away.
 */
function normaliseState(raw) {
  const s = raw && typeof raw === "object" ? raw : {};
  const str = (v) => (typeof v === "string" ? v : "");
  return {
    version: str(s.version),
    available_version: str(s.available_version),
    channel: str(s.channel),
  };
}

/**
 * Validate a drained install request. Returns the request or null.
 *
 * Strict on purpose: this value quits the application. A payload missing its
 * request id or version is a bug somewhere, and acting on it would produce an
 * unexplained restart — the exact failure the step-up exists to make impossible.
 */
function validateInstall(raw) {
  if (!raw || typeof raw !== "object") return null;
  if (typeof raw.request_id !== "string" || !raw.request_id) return null;
  if (typeof raw.version !== "string" || !raw.version) return null;
  return {
    request_id: raw.request_id,
    version: raw.version,
    channel: typeof raw.channel === "string" ? raw.channel : "",
  };
}

function createAppUpdateBridge(deps) {
  const {
    fetchFn,
    getGatewayUrl,
    getSecret,
    getUpdaterState,
    installApproved,
    isGatewayLocal = () => false,
    onError = () => {},
    pollMs = DEFAULT_POLL_MS,
    backoffMs = DEFAULT_BACKOFF_MS,
  } = deps || {};

  if (typeof fetchFn !== "function") throw new Error("createAppUpdateBridge: fetchFn is required");
  if (typeof getGatewayUrl !== "function") throw new Error("createAppUpdateBridge: getGatewayUrl is required");
  if (typeof getSecret !== "function") throw new Error("createAppUpdateBridge: getSecret is required");
  if (typeof getUpdaterState !== "function") throw new Error("createAppUpdateBridge: getUpdaterState is required");
  if (typeof installApproved !== "function") throw new Error("createAppUpdateBridge: installApproved is required");

  let running = false;
  let wakeSleep = null;
  let loopDone = null;
  // Request ids already dispatched. The gateway pops a request on delivery, so
  // this is belt-and-braces against a retry that re-delivers one: an install
  // quits the app, and doing it twice for one approval is not recoverable by
  // the user.
  const dispatched = new Set();

  const report = (err, context) => {
    try {
      onError(err, context);
    } catch {
      /* a log sink must never break the loop */
    }
  };

  function sleep(ms) {
    return new Promise((resolve) => {
      let done = false;
      const finish = () => {
        if (done) return;
        done = true;
        wakeSleep = null;
        resolve();
      };
      wakeSleep = finish;
      const timer = setTimeout(finish, Math.max(0, ms));
      // A poll that runs for the life of the app must never be the reason the
      // process stays alive: unref'd, exactly like auto-update.js's own launch
      // and poll timers. Without this the loop keeps Node's event loop busy
      // forever, which hangs any harness that merely constructed the shell.
      if (timer && typeof timer.unref === "function") timer.unref();
    });
  }

  function headers() {
    const h = { "Content-Type": "application/json" };
    const secret = getSecret();
    if (secret) h["X-Internal-Secret"] = secret;
    return h;
  }

  /**
   * One poll. Returns the validated install request, `null` when there is
   * nothing to install, or `undefined` on a transport error / non-2xx /
   * malformed body — the distinction is what lets the caller back off without
   * inspecting HTTP itself (same contract as the browser channel's drainOnce).
   */
  async function pollOnce() {
    let state;
    try {
      state = normaliseState(await getUpdaterState());
    } catch (err) {
      // A throwing reader still reports presence: see normaliseState.
      report(err, { phase: "updater-state" });
      state = normaliseState(null);
    }
    let res;
    try {
      res = await fetchFn(joinUrl(getGatewayUrl(), BRIDGE_PATH), {
        method: "POST",
        headers: headers(),
        body: JSON.stringify({ state }),
      });
    } catch (err) {
      report(err, { phase: "bridge-fetch" });
      return undefined;
    }
    if (!res || typeof res.status !== "number") {
      report(new Error("bridge: malformed response object"), { phase: "bridge-response" });
      return undefined;
    }
    if (!res.ok) {
      report(new Error(`bridge: HTTP ${res.status}`), { phase: "bridge-status", status: res.status });
      return undefined;
    }
    let body;
    try {
      body = await res.json();
    } catch (err) {
      report(err, { phase: "bridge-parse" });
      return undefined;
    }
    if (!body || typeof body !== "object") {
      report(new Error("bridge: malformed body"), { phase: "bridge-validate" });
      return undefined;
    }
    // An absent / null `install` is the ordinary answer, not a fault.
    if (body.install === null || body.install === undefined) return null;
    const install = validateInstall(body.install);
    if (install === null) {
      report(new Error("bridge: malformed install request"), { phase: "bridge-install-validate" });
      return undefined;
    }
    return install;
  }

  async function dispatchInstall(install) {
    if (dispatched.has(install.request_id)) {
      report(
        new Error(`bridge: install ${install.request_id} already dispatched`),
        { phase: "install-duplicate", id: install.request_id },
      );
      return;
    }
    dispatched.add(install.request_id);
    try {
      await installApproved(install);
    } catch (err) {
      report(err, { phase: "install-dispatch", id: install.request_id });
    }
  }

  async function loop() {
    while (running) {
      // A window attached to a REMOTE gateway must not push the local secret
      // through the tunnel, and a remote gateway's approvals are not about this
      // machine's bundle anyway. Same predicate the browser channel's heartbeat
      // uses, and the same reason.
      if (!isGatewayLocal()) {
        if (running) await sleep(pollMs);
        continue;
      }
      const install = await pollOnce();
      if (!running) break;
      if (install === undefined) {
        await sleep(backoffMs);
        continue;
      }
      if (install !== null) {
        // A failing dispatch must never kill the loop.
        try {
          await dispatchInstall(install);
        } catch (err) {
          report(err, { phase: "install", id: install.request_id });
        }
      }
      if (running) await sleep(pollMs);
    }
  }

  return {
    /** Start polling. Idempotent. */
    start() {
      if (running) return;
      running = true;
      loopDone = loop().catch((err) => {
        report(err, { phase: "loop-crash" });
        running = false;
      });
    },

    /** Poll NOW rather than waiting out the interval (e.g. after a check). */
    poke() {
      if (wakeSleep) wakeSleep();
    },

    /** Stop cleanly, interrupting an in-flight sleep. Resolves once unwound. */
    async stop() {
      running = false;
      if (wakeSleep) wakeSleep();
      const done = loopDone;
      loopDone = null;
      if (done) await done;
    },

    /** Test seam. */
    isRunning: () => running,
  };
}

module.exports = {
  BRIDGE_PATH,
  DEFAULT_POLL_MS,
  DEFAULT_BACKOFF_MS,
  joinUrl,
  normaliseState,
  validateInstall,
  createAppUpdateBridge,
};
